From ba0b371e2f8d60b5034ffa8e7d83570f087b61c4 Mon Sep 17 00:00:00 2001
From: Dan O'Donovan <danielx.o'donovan@intel.com>
Date: Thu, 29 May 2014 17:06:03 +0200
Subject: [PATCH 334/429] gpio: Add platform data init handlers for PCAL9555A
 I2C device registration

The Edison platform contains 4 PCAL9555A I2C GPIO expanders, which are
registered based on entries in the SFI device table.  This patch adds the
platform data init handlers required to register these devices correctly.

Signed-off-by: Dan O'Donovan <danielx.o'donovan@intel.com>
---
 arch/x86/platform/intel-mid/board.c           |  5 ++
 .../platform/intel-mid/device_libs/Makefile   |  1 +
 .../device_libs/platform_pcal9555a.c          | 68 +++++++++++++++++++
 .../device_libs/platform_pcal9555a.h          | 19 ++++++
 4 files changed, 93 insertions(+)
 create mode 100644 arch/x86/platform/intel-mid/device_libs/platform_pcal9555a.c
 create mode 100644 arch/x86/platform/intel-mid/device_libs/platform_pcal9555a.h

diff --git a/arch/x86/platform/intel-mid/board.c b/arch/x86/platform/intel-mid/board.c
index e669ae9ba975..89a470000ee9 100644
--- a/arch/x86/platform/intel-mid/board.c
+++ b/arch/x86/platform/intel-mid/board.c
@@ -73,6 +73,7 @@
 #include "device_libs/platform_mpu3050.h"
 #include "device_libs/platform_tc35876x.h"
 #include "device_libs/platform_bq24261.h"
+#include "device_libs/platform_pcal9555a.h"
 
 /*
  * SPI devices
@@ -99,6 +100,10 @@ struct devs_id __initconst device_ids[] = {
 
 	/* I2C devices*/
 	{"bq24261_charger", SFI_DEV_TYPE_I2C, 1, &bq24261_platform_data, NULL},
+	{"pcal9555a-1", SFI_DEV_TYPE_I2C, 1, &pcal9555a_platform_data, NULL},
+	{"pcal9555a-2", SFI_DEV_TYPE_I2C, 1, &pcal9555a_platform_data, NULL},
+	{"pcal9555a-3", SFI_DEV_TYPE_I2C, 1, &pcal9555a_platform_data, NULL},
+	{"pcal9555a-4", SFI_DEV_TYPE_I2C, 1, &pcal9555a_platform_data, NULL},
 
 	/* SPI devices */
 	{"bma023", SFI_DEV_TYPE_I2C, 1, &no_platform_data, NULL},
diff --git a/arch/x86/platform/intel-mid/device_libs/Makefile b/arch/x86/platform/intel-mid/device_libs/Makefile
index 825a2753d125..a0a82f292642 100644
--- a/arch/x86/platform/intel-mid/device_libs/Makefile
+++ b/arch/x86/platform/intel-mid/device_libs/Makefile
@@ -35,6 +35,7 @@ obj-$(subst m,y,$(CONFIG_GPIO_PCA953X)) += platform_max7315.o
 obj-$(subst m,y,$(CONFIG_SENSORS_MPU3050)) += platform_mpu3050.o
 obj-$(subst m,y,$(CONFIG_GPIO_PCA953X)) += platform_tca6416.o
 obj-$(subst m,y,$(CONFIG_BQ24261_CHARGER)) += platform_bq24261.o
+obj-$(subst m,y,$(CONFIG_GPIO_PCA953X)) += platform_pcal9555a.o
 # SPI Devices
 obj-$(subst m,y,$(CONFIG_SERIAL_MRST_MAX3110)) += platform_max3111.o
 # MISC Devices
diff --git a/arch/x86/platform/intel-mid/device_libs/platform_pcal9555a.c b/arch/x86/platform/intel-mid/device_libs/platform_pcal9555a.c
new file mode 100644
index 000000000000..6625edc7285d
--- /dev/null
+++ b/arch/x86/platform/intel-mid/device_libs/platform_pcal9555a.c
@@ -0,0 +1,68 @@
+/*
+ * platform_pcal9555a.c: pcal9555a platform data initilization file
+ *
+ * (C) Copyright 2014 Intel Corporation
+ * Author: Dan O'Donovan <dan@emutex.com>
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License
+ * as published by the Free Software Foundation; version 2
+ * of the License.
+ */
+
+#include <linux/init.h>
+#include <linux/gpio.h>
+#include <linux/i2c.h>
+#include <linux/i2c/pca953x.h>
+#include <asm/intel-mid.h>
+#include "platform_pcal9555a.h"
+
+
+void __init *pcal9555a_platform_data(void *info)
+{
+	static struct pca953x_platform_data pcal9555a_pdata[PCAL9555A_NUM];
+	static int nr;
+	struct pca953x_platform_data *pcal9555a;
+	struct i2c_board_info *i2c_info = info;
+	int gpio_base, intr;
+	char base_pin_name[SFI_NAME_LEN + 1];
+	char intr_pin_name[SFI_NAME_LEN + 1];
+
+	if (!info) {
+		pr_err("%s: invalid info pointer\n", __func__);
+		return NULL;
+	}
+
+	if (nr >= PCAL9555A_NUM) {
+		pr_err("%s: too many pcal9555a, we only support %d\n",
+			__func__, PCAL9555A_NUM);
+		return NULL;
+	}
+	pcal9555a = &pcal9555a_pdata[nr++];
+
+	/* we have several pcal9555a on the board, we only need load several
+	 * instances of the same pca953x driver to cover them
+	 */
+
+	snprintf(base_pin_name, sizeof(base_pin_name),
+		 "%s_base", i2c_info->type);
+	snprintf(intr_pin_name, sizeof(intr_pin_name),
+		 "%s_int", i2c_info->type);
+
+	strcpy(i2c_info->type, "pcal9555a");
+
+	gpio_base = get_gpio_by_name(base_pin_name);
+	intr = get_gpio_by_name(intr_pin_name);
+
+	if (gpio_base == -1)
+		return NULL;
+	pcal9555a->gpio_base = gpio_base;
+	if (intr != -1) {
+		i2c_info->irq = intr + INTEL_MID_IRQ_OFFSET;
+		pcal9555a->irq_base = gpio_base + INTEL_MID_IRQ_OFFSET;
+	} else {
+		i2c_info->irq = -1;
+		pcal9555a->irq_base = -1;
+	}
+	return pcal9555a;
+}
diff --git a/arch/x86/platform/intel-mid/device_libs/platform_pcal9555a.h b/arch/x86/platform/intel-mid/device_libs/platform_pcal9555a.h
new file mode 100644
index 000000000000..5593db2cc50a
--- /dev/null
+++ b/arch/x86/platform/intel-mid/device_libs/platform_pcal9555a.h
@@ -0,0 +1,19 @@
+/*
+ * platform_pcal9555a.h: pcal9555a platform data header file
+ *
+ * (C) Copyright 2014 Intel Corporation
+ * Author: Dan O'Donovan <dan@emutex.com>
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License
+ * as published by the Free Software Foundation; version 2
+ * of the License.
+ */
+#ifndef _PLATFORM_PCAL9555A_H_
+#define _PLATFORM_PCAL9555A_H_
+
+/* we have multiple pcal9555a on the board ... */
+#define PCAL9555A_NUM 4
+
+extern void __init *pcal9555a_platform_data(void *info) __attribute__((weak));
+#endif
-- 
2.37.3

